### PR TITLE
WIP: add execute to gs.provider

### PR DESCRIPTION
**Pull request by nechaido**
@tshemsedinov @belochub PTAL.

### CODE REVIEW

File: lib/provider.js [hunk 1]
```diff
@@ -79,6 +79,51 @@ class StorageProvider {
   select(query) {
     throw new GSError(errorCodes.NOT_IMPLEMENTED);
   }
+
+  // Check that given user has acces  to execute a function
```
Comment (belochub) [inline lib/provider.js]:
```suggestion
  // Check that given user has access to execute a function
```

File: lib/provider.js [hunk 2]
```diff
@@ -79,6 +79,51 @@ class StorageProvider {
...
+  //   action - string, name of an Action to execute
```
Comment (belochub) [inline lib/provider.js]: I think you can use the new documentation comment format to reduce the number of changes when refactoring all of the comments in the project.

File: lib/errors.js [hunk 3]
```diff
@@ -5,10 +5,12 @@ const codes = {
   NOT_FOUND: 1001,
   INVALID_SCHEMA: 1002,
   INVALID_CATEGORY_TYPE: 1003,
+  NO_SUCH_ACTION: 1004,
 };
 
 const defaultMessages = {
   [codes.NOT_IMPLEMENTED]: 'Not implemented',
+  [codes.NO_SUCH_ACTION]: 'No such action',
```
Comment (belochub) [inline lib/errors.js]: I am against adding a default message for this case, I think you should always specify the action that is missing.

File: lib/provider.js [hunk 4]
```diff
@@ -79,6 +79,51 @@ class StorageProvider {
   select(query) {
     throw new GSError(errorCodes.NOT_IMPLEMENTED);
   }
+
+  // Check that given user has access  to execute a function
```
Comment (belochub) [inline lib/provider.js]:
```suggestion
  // Check that given user has access to execute a function
```

File: lib/provider.js [hunk 5]
```diff
@@ -79,6 +79,51 @@ class StorageProvider {
...
+      callback(new GSError(errorCodes.NO_SUCH_ACTION));
+      return;
+    }
+    this.checkPermission(name, category, executor, args, (error, permitted) => {
+      if (error) {
+        callback(error);
+        return;
+      }
+      if (!permitted) {
+        callback(new GSError(errorCodes.NOT_AUTHORISED));
```
Comment (belochub) [inline lib/provider.js]: You haven't defined this error code.

File: lib/provider.js [hunk 6]
```diff
@@ -79,6 +79,51 @@ class StorageProvider {
...
+        return;
+      }
+      const valid = this.schema.validateForm(form, args);
+      if (!valid) {
+        callback(new GSError(errorCodes.INVALID_SIGNATURE));
```
Comment (belochub) [inline lib/provider.js]: Ditto.

File: lib/provider.js [hunk 7]
```diff
@@ -79,6 +79,51 @@ class StorageProvider {
...
+  }
+
+  // Execute Action
+  //   action - <string>, name of an Action to execute
+  //   executor - number>, id of user that executes an Action
+  //   args - <Object>, object that contains arguments of an action
+  //   callback - <function>
+  //     error - <Error>
+  //     message - <string>
+  execute(name, category, executor, args, callback) {
```
Comment (tshemsedinov) [inline lib/provider.js]:
```suggestion
  execute(category, name, executor, args, callback) {
```

File: lib/provider.js [hunk 8]
```diff
@@ -79,6 +79,51 @@ class StorageProvider {
...
+    const form = `${category}.${name}`;
+    const action = this.schema.action.get(form);
+    if (!action) {
+      callback(new GSError(errorCodes.NO_SUCH_ACTION));
+      return;
+    }
+    this.checkPermission(name, category, executor, args, (error, permitted) => {
```
Comment (tshemsedinov) [inline lib/provider.js]: Ditto

File: lib/provider.js [hunk 9]
```diff
@@ -79,6 +79,51 @@ class StorageProvider {
...
+        callback(new GSError(errorCodes.NOT_AUTHORISED));
+        return;
+      }
+      const valid = this.schema.validateForm(form, args);
+      if (!valid) {
+        callback(new GSError(errorCodes.INVALID_SIGNATURE));
+        return;
+      }
+      // eslint-disable-next-line new-cap
+      action.Execute(args, callback);
```
Comment (tshemsedinov) [inline lib/provider.js]:
```js
const { Execute: fn } = action;
fn(args, callback);
```

Comment (tshemsedinov) [reply lib/provider.js]: remove eslint disable-next-line

File: lib/provider.js [hunk 10]
```diff
@@ -79,6 +79,51 @@ class StorageProvider {
...
+  //     error - <Error>
+  //     permission - <boolean>
+  // eslint-disable-next-line no-unused-vars
+  checkPermission(action, executor, args, callback) {
+    throw new GSError(errorCodes.NOT_IMPLEMENTED);
+  }
+
+  // Execute Action
+  //   action - <string>, name of an Action to execute
+  //   executor - number>, id of user that executes an Action
```
Comment (tshemsedinov) [inline lib/provider.js]:
```suggestion
  //   userId - <common.Uint64>, id of user that executes an Action
```

File: lib/provider.js [hunk 11]
```diff
@@ -79,6 +79,51 @@ class StorageProvider {
...
+  //   callback - <function>
+  //     error - <Error>
+  //     permission - <boolean>
+  // eslint-disable-next-line no-unused-vars
+  checkPermission(action, executor, args, callback) {
+    throw new GSError(errorCodes.NOT_IMPLEMENTED);
+  }
+
+  // Execute Action
+  //   action - <string>, name of an Action to execute
```
Comment (tshemsedinov) [inline lib/provider.js]: Add `categody` argument

File: lib/provider.js [hunk 12]
```diff
@@ -83,6 +83,53 @@ class StorageProvider {
...
+    this.checkPermission(category, name, userId, args, (error, permitted) => {
+      if (error) {
+        callback(error);
+        return;
+      }
+      if (!permitted) {
+        callback(new GSError(errorCodes.NOT_AUTHORISED));
+        return;
+      }
+      const valid = this.schema.gs.validateForm(form, args);
```
Comment (belochub) [inline lib/provider.js]:
```suggestion
      const valid = this.gs.schema.validateForm(form, args);
```

File: lib/provider.js [hunk 13]
```diff
@@ -83,6 +83,53 @@ class StorageProvider {
...
+      callback(new GSError(errorCodes.NO_SUCH_ACTION));
+      return;
+    }
+    this.checkPermission(category, name, userId, args, (error, permitted) => {
+      if (error) {
+        callback(error);
+        return;
+      }
+      if (!permitted) {
+        callback(new GSError(errorCodes.NOT_AUTHORISED));
```
Comment (belochub) [inline lib/provider.js]: There is no such error code defined, probably a typo:
```suggestion
        callback(new GSError(errorCodes.NOT_AUTHORIZED));
```

File: lib/provider.js [hunk 14]
```diff
@@ -83,6 +83,53 @@ class StorageProvider {
...
+  //   userId - <common.Uint64>, id of user that executes an Action
+  //   args - <Object>, object that contains arguments of an action
+  //   callback - <function>
+  //     error - <Error>
+  //     message - <string>
+  execute(category, name, userId, args, callback) {
+    const form = `${category}.${name}`;
+    const action = this.gs.schema.action.get(form);
+    if (!action) {
+      callback(new GSError(errorCodes.NO_SUCH_ACTION));
```
Comment (belochub) [inline lib/provider.js]: You haven't added an error message as requested in https://github.com/metarhia/globalstorage/pull/259#discussion_r230128561, even though you have resolved the discussion there.

File: lib/provider.js (hunk 13 above)
Comment (belochub) [reply lib/provider.js]: Also, this is missing an error message.

File: lib/provider.js [hunk 15]
```diff
@@ -83,6 +83,53 @@ class StorageProvider {
...
+        callback(error);
+        return;
+      }
+      if (!permitted) {
+        callback(new GSError(errorCodes.NOT_AUTHORISED));
+        return;
+      }
+      const valid = this.schema.gs.validateForm(form, args);
+      if (!valid) {
+        callback(new GSError(errorCodes.INVALID_SIGNATURE));
```
Comment (belochub) [inline lib/provider.js]: The error message is missing here as well.

File: lib/provider.js (hunk 12 above)
Comment (belochub) [reply lib/provider.js]: Also, I think, we should receive the errors from `validateForm` here, and pass them into the callback as well (as part of an `INVALID_SIGNATURE` error's message).

File: lib/provider.js [hunk 16]
```diff
@@ -83,6 +83,62 @@ class StorageProvider {
...
+  //   category - <string>, name of a category that the action is defined upon
+  //   name - <string>, name of an Action to execute
+  //   userId - <common.Uint64>, id of user that executes an Action
+  //   args - <Object>, object that contains arguments of an action
+  //   callback - <function>
+  //     error - <Error>
+  //     message - <string>
+  execute(category, name, userId, args, callback) {
+    const form = `${category}.${name}`;
+    const action = this.gs.schema.action.get(form);
```
Comment (lundibundi) [inline lib/provider.js]:
```suggestion
    const action = this.gs.schema.actions.get(form);
```
?

File: lib/provider.js [hunk 17]
```diff
@@ -83,6 +83,62 @@ class StorageProvider {
...
+    if (!action) {
+      callback(new GSError(
+        errorCodes.NO_SUCH_ACTION,
+        `There is no such action as '${name}' in '${category}' category`
```
Comment (lundibundi) [inline lib/provider.js]:
```suggestion
        `There is no such action '${name}' in '${category}' category`
```
?

File: lib/provider.js [hunk 18]
```diff
@@ -83,6 +83,62 @@ class StorageProvider {
...
+        return;
+      }
+      if (!permitted) {
+        callback(new GSError(
+          errorCodes.NOT_AUTHORIZED,
+          `User '${userId.toString()}' is not authorized to execute '${form}'`
+        ));
+        return;
+      }
+      const validationError = this.schema.gs.validateForm(form, args);
```
Comment (lundibundi) [inline lib/provider.js]:
```suggestion
      const validationError = this.gs.schema.validateForm(form, args);
```
?

File: lib/provider.js [hunk 19]
```diff
@@ -83,6 +83,62 @@ class StorageProvider {
...
+  }
+
+  // Execute Action
+  //   category - <string>, name of a category that the action is defined upon
+  //   name - <string>, name of an Action to execute
+  //   userId - <common.Uint64>, id of user that executes an Action
+  //   args - <Object>, object that contains arguments of an action
+  //   callback - <function>
+  //     error - <Error>
+  //     message - <string>
```
[the remark this answers is not in the document]
Comment (belochub) [reply lib/provider.js]: @lundibundi we are following the CQS principle, thus there is no result of the execution.

File: lib/provider.js [hunk 20]
```diff
@@ -83,6 +83,62 @@ class StorageProvider {
   select(category, query) {
     throw new GSError(errorCodes.NOT_IMPLEMENTED);
   }
+
+  // Check that given user has access to execute a function
+  //   category - <string>, name of a category that the action is defined upon
+  //   name - <string>, name of an Action to execute
+  //   userId - <common.Uint64>, id of user that executes an Action
+  //   args - <Object>, object that contains arguments of an action
```
Comment (belochub) [inline lib/provider.js]:
```suggestion
  //   args - <Object>, object that contains arguments of the Action
```

File: lib/provider.js [hunk 21]
```diff
@@ -83,6 +83,62 @@ class StorageProvider {
   select(category, query) {
     throw new GSError(errorCodes.NOT_IMPLEMENTED);
   }
+
+  // Check that given user has access to execute a function
+  //   category - <string>, name of a category that the action is defined upon
+  //   name - <string>, name of an Action to execute
+  //   userId - <common.Uint64>, id of user that executes an Action
```
Comment (belochub) [inline lib/provider.js]:
```suggestion
  //   userId - <common.Uint64>, id of user that executes the Action
```

File: lib/provider.js [hunk 22]
```diff
@@ -83,6 +83,62 @@ class StorageProvider {
...
+  //     permission - <boolean>
+  // eslint-disable-next-line no-unused-vars
+  checkPermission(category, name, userId, args, callback) {
+    throw new GSError(errorCodes.NOT_IMPLEMENTED);
+  }
+
+  // Execute Action
+  //   category - <string>, name of a category that the action is defined upon
+  //   name - <string>, name of an Action to execute
+  //   userId - <common.Uint64>, id of user that executes an Action
```
Comment (belochub) [inline lib/provider.js]:
```suggestion
  //   userId - <common.Uint64>, id of user that executes the Action
```

File: lib/provider.js [hunk 23]
```diff
@@ -83,6 +83,62 @@ class StorageProvider {
...
+  //   args - <Object>, object that contains arguments of an action
```
Comment (belochub) [inline lib/provider.js]:
```suggestion
  //   args - <Object>, object that contains arguments of the Action
```

File: lib/provider.js [hunk 24]
```diff
@@ -83,6 +88,195 @@ class StorageProvider {
...
+          }
+
+          return true;
+        });
+
+        if (permission) {
+          callback(null, true);
+        }
+      });
+      callback(null, false);
```
Comment (belochub) [inline lib/provider.js]: You seem to be calling `callback` multiple times in this function.

File: lib/provider.js [hunk 25]
```diff
@@ -83,6 +88,216 @@ class StorageProvider {
...
+
+      for (const { Permissions } of user.Roles) {
+        const permission = Permissions.find(permission => {
+          if (actionId && !permission.Actions.find(
+            action => UInt64.cmp(action, actionId) === 0
+          )) {
+            return false;
+          }
+
+          if (accessFlag && permission.Access.get(accessFlag)) {
```
Comment (lundibundi) [inline lib/provider.js]:
```suggestion
          if (accessFlag & permission.Access.get(accessFlag) === 0) {
```
?

File: lib/provider.js [hunk 26]
```diff
@@ -83,6 +88,216 @@ class StorageProvider {
...
+            return false;
+          }
+
+          if (permission.Category.equal(categoryId)) {
```
Comment (lundibundi) [inline lib/provider.js]:
```suggestion
          if (!permission.Category.equal(categoryId)) {
```
?

File: lib/provider.js [hunk 27]
```diff
@@ -83,6 +88,216 @@ class StorageProvider {
...
+            return false;
+          }
+
+          if (UInt64.cmp(permission.Target, args.Identifier) === 0) {
```
Comment (lundibundi) [inline lib/provider.js]:
```suggestion
          if (permission.Target &&
            UInt64.cmp(permission.Target, args.Identifier) === 0) {
```

File: lib/provider.js [hunk 28]
```diff
@@ -83,6 +85,227 @@ class StorageProvider {
...
+        new GSError(
+          errorCodes.INVALID_SCHEMA, `Undefined action '${name}'`
+        )
+      );
+      return;
+    }
+
+    const actionId = this.gs.permissionCache.Action.get(name).Id;
+    const categoryId = this.gs.permissionCache.Category.get(category).Id;
+    if (!action) {
```
Comment (lundibundi) [inline lib/provider.js]: Not needed?

**Comment by belochub** [comment]
Closed in favor of https://github.com/metarhia/globalstorage/pull/318.

**Pull request closed without merging**